external/error_report/error_report.c

```diff
@@ -0,0 +1,205 @@
+
+_rt_error_data_s g_error_report[CONFIG_ERROR_REPORT_NENTRIES];
+typedef enum {
+	ERRSTATE_UNINITIALIZED = -1,
+	ERRSTATE_INITIALIZED,
+	ERRSTATE_MEMUNDERFLOW,
+	ERRSTATE_BUFFER_EXCEEDED,
+} err_fsm_e;
+
+typedef struct {
+	uint16_t rear;
+	uint16_t front;
+	int8_t q_pending;
+	pthread_mutex_t err_mutex;
+	err_fsm_e fsm;
+} g_error_info_s;
+
+static g_error_info_s g_err_info;
+
+static unsigned long prv_fetch_taskaddr(int pid)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_MAX_TASKS; i++) {
+		if ((g_pidhash[i].tcb)->pid == pid) {
+			entry_t e;
+			e = g_pidhash[i].tcb->entry;
+			return (unsigned long)e.pthread;
+		}
+	}
+
+	return 0;
+}
+
+_rt_err_status_e error_report_init(void)
+{
+	pthread_mutex_init(&g_err_info.err_mutex, NULL);
+	g_err_info.fsm = ERRSTATE_INITIALIZED;
+	return ERR_SUCCESS;
+}
+
+_rt_err_status_e error_report_deinit(void)
+{
+	pthread_mutex_destroy(&g_err_info.err_mutex);
+
+	return ERR_SUCCESS;
+}
+
+_rt_error_data_s *error_report_data_create(int module_id, int error_code, uint32_t pc_value)
+{
+	struct timeval ts;
+	gettimeofday(&ts, NULL);
+	_rt_error_data_s *ret = NULL;
+
+	pthread_mutex_lock(&g_err_info.err_mutex);
+
+	ret = (_rt_error_data_s *) g_error_report + g_err_info.rear;
+	ret->timestamp.tv_sec = ts.tv_sec;
+	ret->timestamp.tv_usec = ts.tv_usec;
+	ret->module_id = module_id;
+	ret->error_code = error_code;
+	ret->pc_value = pc_value;
+	rt_errdebug("pc_value: %08x\n", pc_value);
+	ret->task_addr = prv_fetch_taskaddr(getpid());
+	rt_errdebug("task_addr: %08x\n", ret->task_addr);
+	if (g_err_info.fsm == ERRSTATE_MEMUNDERFLOW) {
+		g_err_info.fsm = ERRSTATE_INITIALIZED;
+	}
+	g_err_info.rear = (g_err_info.rear + 1) % CONFIG_ERROR_REPORT_NENTRIES;
+	if (g_err_info.q_pending < CONFIG_ERROR_REPORT_NENTRIES) {
+		g_err_info.q_pending++;
+	}
+	pthread_mutex_unlock(&g_err_info.err_mutex);
+	return ret;
+}
+
+int error_report_data_read(char *readbuf)
+{
+	int nbytes = 0;
+	int i;
+	int nentries = g_err_info.q_pending;
+	pthread_mutex_lock(&g_err_info.err_mutex);
+	if (g_err_info.q_pending < CONFIG_ERROR_REPORT_NENTRIES) {
+		g_err_info.front = 0;
+	} else {
+		g_err_info.front = g_err_info.rear;
+	}
+	if (readbuf != NULL) {
+		if (!nentries) {
+			g_err_info.fsm = ERRSTATE_MEMUNDERFLOW;
+		} else {
+			while (nentries) {
+				if (nbytes + sizeof(_rt_error_data_s) >= ERR_BUFLEN) {
+					g_err_info.fsm = ERRSTATE_BUFFER_EXCEEDED;
+					goto err_read_done;
+				}
+				memcpy(readbuf + nbytes, (char *)(g_error_report + g_err_info.front), sizeof(_rt_error_data_s));
+				nbytes += sizeof(_rt_error_data_s);
+				g_err_info.front = (g_err_info.front + 1) % CONFIG_ERROR_REPORT_NENTRIES;
+				nentries--;
+			}
+			*(readbuf + nbytes) = '\0';
+		}
+	}
+
+err_read_done:
+	rt_errdebug("Report in Hex: ");
+	for (i = 0; i < nbytes; i++) {
+		rt_errdebug("%02x ", readbuf[i]);
+	}
+	rt_errdebug("\n");
+
+	pthread_mutex_unlock(&g_err_info.err_mutex);
+	return nbytes;
+}
+
+int error_report_send(const char *ep, char *readbuf, int readbuf_len)
```

Comment:
In addition to this, what about adding an API to export the serialized output (e.g., memory devices such as USB) ?

Reply:
Thank you again. Yes, it would be a good idea to export the logs to persistent storage. However, this feature add-on will be scheduled for a future commit.